File: cp6.web/src/utils/wmsHub.ts
import * as signalR from '@microsoft/signalr'

/**
 * WMS SignalR Hub クライアント（独立 connection）
 *
 * /hubs/wms に接続して以下イベントを購読：
 *   StockChanged / InboundReceived / OutboundShipped / StockTakeCompleted
 *
 * 既存の /hubs/notify（汎用）/ /hubs/mes と完全独立 — Hub ごとに connection 分離。
 */

export interface StockChangedPayload {
  txnNo: string
  txnType: 'IN' | 'OUT' | 'MOVE' | 'ADJ' | 'RSV' | 'UNRSV'
  txnAt: string
  warehouseCd: string
  locationCd: string
  productCd: string
  lotNo: string
  qty: number
  relatedNo?: string
  operatorCd?: string
}

export interface InboundReceivedPayload {
  receiptNo: string
  warehouseCd: string
  at: string
}

export interface OutboundShippedPayload {
  outboundNo: string
  packageNo?: string
  at: string
}

export interface StockTakeCompletedPayload {
  stockTakeNo: string
  diffLines: number
  at: string
}

let wmsConn: signalR.HubConnection | null = null

export function getWmsConnection(): signalR.HubConnection {
  if (!wmsConn) {
    wmsConn = new signalR.HubConnectionBuilder()
      .withUrl('/hubs/wms')
      .withAutomaticReconnect([0, 2000, 5000, 10000, 30000])
      .configureLogging(signalR.LogLevel.Warning)
      .build()
  }
  return wmsConn
}

export async function startWmsConnection() {
  const c = getWmsConnection()
  if (c.state === signalR.HubConnectionState.Disconnected) {
    try {
      await c.start()
      console.log('[WMS-Hub] Connected')
    } catch (err) {
      console.warn('[WMS-Hub] Connection failed, will retry:', err)
    }
  }
  return c
}

export function stopWmsConnection() {
  if (wmsConn) {
    wmsConn.stop()
    wmsConn = null
  }
}

/** 倉庫別 購読登録（接続中のみ即時、それ以外は内部キュー無し ─ 再接続時に再購読要） */
export async function subscribeWarehouse(warehouseCd: string) {
  const c = getWmsConnection()
  if (c.state === signalR.HubConnectionState.Connected) {
    await c.invoke('SubscribeWarehouse', warehouseCd)
  }
}

export async function subscribeProduct(productCd: string) {
  const c = getWmsConnection()
  if (c.state === signalR.HubConnectionState.Connected) {
    await c.invoke('SubscribeProduct', productCd)
  }
}

File: CP6.WebApi/Hubs/WmsHub.cs
using Microsoft.AspNetCore.SignalR;

namespace CP6.WebApi.Hubs;

/// <summary>
/// WMS SignalR Hub — 倉庫リアルタイムイベント
/// </summary>
/// <remarks>
/// 推送事件：
/// - StockChanged：在庫変動（IN/OUT/MOVE/ADJ/RSV/UNRSV）
/// - InboundReceived：入庫実績確定
/// - OutboundShipped：出庫確定 + 梱包NO
/// - StockTakeCompleted：棚卸完了
///
/// クライアント接続：
///   const conn = new HubConnectionBuilder().withUrl('/hubs/wms').build()
///   conn.on('StockChanged', payload => { ... })
///
/// 購読グループ：
///   - wh:{warehouseCd}      倉庫別の絞り込み
///   - product:{productCd}    製品別の絞り込み
/// </remarks>
public class WmsHub : Hub
{
    private readonly ILogger<WmsHub> _logger;
    public WmsHub(ILogger<WmsHub> logger) => _logger = logger;

    public override Task OnConnectedAsync()
    {
        _logger.LogInformation("WMS Hub 接続: {ConnectionId}", Context.ConnectionId);
        return base.OnConnectedAsync();
    }

    public override Task OnDisconnectedAsync(Exception? exception)
    {
        _logger.LogInformation("WMS Hub 切断: {ConnectionId}", Context.ConnectionId);
        return base.OnDisconnectedAsync(exception);
    }

    /// <summary>クライアントが特定倉庫の更新を購読</summary>
    public Task SubscribeWarehouse(string warehouseCd)
        => Groups.AddToGroupAsync(Context.ConnectionId, $"wh:{warehouseCd}");

    public Task UnsubscribeWarehouse(string warehouseCd)
        => Groups.RemoveFromGroupAsync(Context.ConnectionId, $"wh:{warehouseCd}");

    /// <summary>クライアントが特定製品の更新を購読</summary>
    public Task SubscribeProduct(string productCd)
        => Groups.AddToGroupAsync(Context.ConnectionId, $"product:{productCd}");

    public Task UnsubscribeProduct(string productCd)
        => Groups.RemoveFromGroupAsync(Context.ConnectionId, $"product:{productCd}");
}
